### PR TITLE
Improve support for submodules in code_style.py

### PR DESCRIPTION
**Pull request by Harry-Ramsey**
This commit improves support for submodules within code_style.py. A new function get_submodule_hash retrieves the submodule hash at the time of a commit allowing an appropriate diff to be seen.

Issue: [24](https://github.com/Mbed-TLS/mbedtls-framework/issues/24)

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** not required because: a script change that fixes incorrect behaviour.
- [x] **development PR** provided.
- [x] **framework PR** not required.
- [x] **3.6 PR** #9696.
- [x] **2.28 PR** not required because: framework submodule did not exist.
- **tests**  provided | not required because: 

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

Help make review efficient:
* Multiple simple commits
  - please structure your PR into a series of small commits, each of which does one thing
* Avoid force-push
  - please do not force-push to update your PR - just add new commit(s)
* See our [Guidelines for Contributors](https://mbed-tls.readthedocs.io/en/latest/reviews/review-for-contributors/) for more details about the review process.

### CODE REVIEW

File: scripts/code_style.py
Comment (gilles-peskine-arm) [inline scripts/code_style.py:149]: Minor: this is a slightly fragile way of extracting the sha. It could fail if we add a submodule whose name just happens to be 40 hex digits, or if git ever starts to emit longer shas. The format of `git ls-tree` defines the sha as the third whitespace-separated field, so we can just do
```
return output.split()[2]
```

File: scripts/code_style.py
Comment (gilles-peskine-arm) [inline scripts/code_style.py:154]: Returning the empty string here doesn't actually work. The caller creates a command line with an empty argument, i.e.
`["git", "-C", "framework", "log", "", "HEAD"`, which causes Git to complain
```
fatal: bad revision ''
```

You can try this with e.g. `code_style.py --since mbedtls-3.5.0` from a directory on `development` or 3.6.

It's not important to support this case. `code_style.py --since …` is mostly useful to compare the branch you're working on with the point where it forked, which is a recent commit. If we really wanted to support going back to when the submodule didn't exist, then we should treat all the files in the submodule as in scope. But, again, we don't actually care about this.

File: scripts/code_style.py
Comment (gilles-peskine-arm) [inline scripts/code_style.py:142]: Pylint says: please document this function.

Though as a human, I think its purpose is clear enough. But a one-line description wouldn't hurt.

One thing that is not clear without looking at the code is the behavior if the submodule doesn't exist in the commit. For this particular script, we don't care (see my comment below), but if the function is reused in the future, we might care.

**Review by gilles-peskine-arm — APPROVED** [review]
Thanks for doing this! Looks good to me (I left some optional comments), though the CI isn't happy.

To get DCO to pass, add a signoff line by running `git commit --amend -s`. See team wiki or chat for how to set this up automatically.

**Comment by Harry-Ramsey** [comment]
Thank you for the feedback. I have implemented these changes now. 
Unfortunately I have had to force push to update the commit message and it is now signed, hopefully this does not make reviewing more difficult.

File: scripts/code_style.py
Comment (gilles-peskine-arm) [inline scripts/code_style.py:101]: The `m` from `main` somehow disappeared in the rebase.

File: scripts/code_style.py
Comment (gilles-peskine-arm) [inline scripts/code_style.py:143]: Not a blocker because we don't use a documentation processing tool, but [PEP 8](https://peps.python.org/pep-0008/#documentation-strings) says to write the first line of the docstring immediately after the opening `"""`.

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Comment by gilles-peskine-arm** [comment]
> Will this require a backport to 3.6?

Yes, since 3.6 has the same script and has the framework. Please make a backport once this pull request has two approvals.

File: scripts/code_style.py
Comment (gilles-peskine-arm) [reply scripts/code_style.py]: For future reference (it doesn't matter for this time), in this project, the author of a review comment is the one who's supposed to mark it as resolved, when they're happy about the resolution. As a pull request author, if you want to keep track of which comments you've handled, we use the :rocket: reaction (you don't need to do it if you don't find it useful).

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

**Comment by eleuzi01** [comment]
This looks like it's almost good to go, just needs 3.6 backport

**Comment by Harry-Ramsey** [comment]
Thanks for the review, backport can be found here: #9696.

**Pull request merged**
